text/0000-multi-api-lifecycle-descriptor.md

```diff
@@ -0,0 +1,176 @@
+# Meta
+[meta]: #meta
+- Name: Lifecycle Descriptor with Multiple APIs
+- Start Date: 2020-07-02
+- Author(s): ekcasey
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC-0011](https://github.com/buildpacks/rfcs/blob/main/text/0011-lifecycle-descriptor.md)
+
+# Summary
+[summary]: #summary
+
+This RFC proposed three levels of API support: `deprecated`, `supported`, and `experimental`. Supported APIs will be expressed as arrays.
+
+This RFC introduces a new lifecycle descriptor file format, and an analogous label that should be applied to builder and lifecycle images.
+
+# Motivation
+[motivation]: #motivation
+
+Now that the lifecycle may simultaneously support multiple APIs, the current lifecycle descriptor format, which lists exactly one platform and one buildpack API, is no longer sufficient.
+
+In addition, we should have a mechanism for deprecating an entire API, so that buildpack and platform authors are not caught off guard when their platform or buildpack no longer works with a newer lifecycle. We likely will not choose to deprecate any APIs in the near future but building API deprecation into any new design prepares us well for the future.
+
+Experimental APIs will give platform and buildpack authors an opportunity to try out unreleased APIs and give us feedback. This will be valuable as we prepare for the `1.0` versions of the APIs. We want to get the design of the `1.0` APIs as close to "right" as possible because `1.n` APIs will not permit breaking changes. Feedback will help us improve the design before we finalize it. The quality of feedback we get from user about their real life usage of experimental APIs will help us avoid problems we may miss when approving RFCs.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Lifecycle Descriptor
+The following is an example `lifecycle.toml` following the proposed schema:
+```toml
+[apis]
+[apis.buildpack]
+  deprecated = ["1"]
+  supported = ["1.2", "2.1"]
+  experimental = ["3.0"]
+[apis.platform]
+  deprecated = ["0.4"]
+  supported = ["0.4", "0.5", "1.3"]
+  experimental = ["1.4", "2.0"]
+
+[lifecycle]
+  version = "0.9.0"
+```
+* `supproted` contains an array of support API versions:
+  * for version `1.0+`, version `x.n` implies support for [`x.0`,`x.n`]
+  * `supported` should be a superset of `deprecated`
```

Comment:
It is because `We likely will not choose to deprecate any APIs in the near future but building API deprecation into any new design prepares us well for the future.`, right?

Moreover, according to your example, under `apis.buildpack`, if `1` is deprecated, `1` should be also `supported` (and not only `1.2`), right?

Reply:
In this example the deprecated `1` refers to the entire `1.x` line:
- `1.2` in `supported` means `1.0-1.2` are supported
- `1` in `deprecated` means all supported `1.x` APIs (i.e. `1.0-1.2`) are deprecated

Therefore `supported` is a superset of deprecated semantically, even though the strings are not literally the same.

Reply:
I know that `1` refers to the entire `1.x` but I didn't know that if `1.2` is `supported` it means that `1.0-1.2` are all `supported`.
A few questions about it:
1. Did you write that if `x.y` is `supported`, it means that `x.0-x.y` are `supported`?
2. I'm afraid that there might be a new version `1.3` and then it will be part of the `deprecated` and not part of the `supported`. Can it happen?
3. Does `supported` have to be a superset of `deprecated` because `We likely will not choose to deprecate any APIs in the near future`? Otherwise I don't understand why are we doing that.

Reply:
> 1. Did you write that if x.y is supported, it means that x.0-x.y are supported?

Yes. That is what  
>  * for version `1.0+`, version `x.n` implies support for [`x.0`,`x.n`]`

is attempting to convey.

> 2. I'm afraid that there might be a new version 1.3 and then it will be part of the deprecated and not part of the supported. Can it happen?

Because `1.2` is the max supported and deprecated must be a subset of supported, `1` in deprecated implies `1.0 - 1.2`.

> 3. Does supported have to be a superset of deprecated because We likely will not choose to deprecate any APIs in the near future? Otherwise I don't understand why are we doing that.

In the original RFC they were disjoint sets, but @sclevine made the valid point that deprecated APIs are in fact "supported" in that they will run successfully. Now `deprecated` is a special type of support, which I think makes sense.